dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/EventTrackerConverterService.java

```diff
@@ -190,12 +190,19 @@ private List<EventDataValue> getProgramStageInstanceDataValues( TrackerPreheat p
             return eventDataValues;
         }
 
-        Set<String> dataElements = event.getDataValues()
+        Set<MetadataIdentifier> dataElements = event.getDataValues()
             .stream()
             .map( DataValue::getDataElement )
             .collect( Collectors.toSet() );
         for ( EventDataValue eventDataValue : programStageInstance.getEventDataValues() )
         {
+            // TODO() what is
+            // org.hisp.dhis.eventdatavalue.EventDataValue.getDataElement()
+            // String? always a UID?
+            // we compare to data values from the payload which can be in any
+            // idScheme
+            // can we first get the DataValue out of the preheat, so we can
+            // compare using UIDs?
             if ( !dataElements.contains( eventDataValue.getDataElement() ) )
```

Comment:
@stian-sandvold and @enricocolasante I need your input on this

org.hisp.dhis.eventdatavalue.EventDataValue.getDataElement() is of type String

Its not clear to me what it represents, I assume a UID. Is that a safe assumption?
If we can safely assume its a UID, I would need to fetch the DataElement from the preheat. Otherwise, I cannot compare it to a DataElement in the payload which can be of any idScheme.

This means that a DataElement in the payload might not be in the preheat => does not exist.

Reply:
eventdatavalue.getDataElement would be a String of type UID. We had an issue where it could be stored with any identifier, but I dont see the point of creating any extra complicated code for that when it should be fixed as its own thing :)

Reply:
@stian-sandvold will know better but I cannot see how dataElement here could be something different from UID.
If at any time we saved this value as any idScheme that was used in the importer it seems to me that we lost this information now